README
============================================================================
WildCAT
Copyright (c) 2004-2005 Universite de Nantes (LINA)
Copyright (c) 2005-2006 France Telecom

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License version 2.1 as published by the Free Software Foundation.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Contact: Pierre-Charles David <pcdavid@gmail.com>
============================================================================

What is WildCAT?
----------------

WildCAT is a Java toolkit/framework to ease the creation of
context-aware applications. It provides an extensible set of sensors
to gather various kinds of contextual data, which can then be
organized into a structured and dynamic model representing an
application's execution context. Client application can query this
model and register listener to be notified automatically when specific
conditions occur.

WildCAT is a free software distributed under the terms of the GNU
Lesser General Public license and kinldy hosted by ObjectWeb at
http://wildcat.objectweb.org/

Getting the source
------------------

The latest releases are available from the project's website at
http//wildcat.objectweb.org/ 

You can get anonymous access to the very latest state of the source
using CVS at :pserver:anonymous@cvs.forge.objectweb.org/cvsroot/wildcat

Building
--------

WildCAT requires Java 5 (JDK 1.5) to build and run. The build process
uses Ant [1] (tested with 1.6.5), and requires the ANTLR parser generator
tool [2] and the JUnit framework [3]. Both are included in the externals/
folder, but must be made available to Ant. You can either:

1. Install the JARs system-wide (if you have the permissions) in Ant's
   optional tasks directory:

  % cp externals/antlr-2.7.6.jar $ANT_HOME/lib
  % cp externals/junit-4.1.jar $ANT_HOME/lib

2. Install the JARs user-wide:

  % cp externals/antlr-2.7.6.jar $HOME/.ant/lib
  % cp externals/junit-4.1.jar $HOME/.ant/lib

3. Tell Ant where to find the JARs using command-line arguments each
   time you invoke it, e.g.

  % ant -lib externals/antlr-2.7.6.jar:externals/junit-4.1.jar test



Building with Eclipse
---------------------

The project is setup for Eclipse [4], and you can install the ANTLR
Eclipse plugin [5] which provides an editor for ANTLR grammar files
and can be used to regenerate the parser from inside Eclipse
(right-click on the grammar file and select "Compile ANTLR grammar").
The automatic parser generation is disabled by default, as the code
generated by ANTLR triggers a lot of (innocuous) warnings which clutter
the Problems view. These warnings can be disabled with some manual
editing: 

1. Navigate to the org.objectweb.wildcat.standard package in the main
   source folder (src/).
2. Select all the generated Java files (Expression*.java) and do
   "Source > Organize imports."
3. Optionally, do also a "Source > Format" to make it more readable.
4. Annotate the lexer and parser classes (ExpressionsLexer and
   ExpressionsReader) with:
   
     @SuppressWarnings({"unchecked", "unused"})
     
   This tells the compiler not to worry that the generated code uses
   pre-Java 5 collections and declares lots of variables it does not
   actually use.

Authors
-------
Main and currently only author:
- Pierre-Charles David <pcdavid@gmail.com>

References
----------

[1] http://ant.apache.org/
[2] http://antlr.org/
[3] http://junit.org/
[4] http://www.eclipse.org/
[5] http://antlreclipse.sourceforge.net/

HISTORY.txt

--------------------------------------------------------------------
History File
--------------------------------------------------------------------

2003/12/19:
 - first version.

2004/01/06:
 - updated copyright footnote.

2004/01/09:
 - removed uneeded call to disableStyleSheet() javascript function in
   browsable stylesheet (body tag).

 - removed disableStyleSheet() function in js/rolls.js (the stylesheet
   is disabled for netscape 4.x thanks to the id="stylesheet"
   statement that is not recognized by this old browser.

2004/01/13:
 - Remplaced Basic program by a Java program in the sample file!

 - Created if needed html and cgi-bin directories in the build.xml (if
   these directories are ampty in CVS, a checkout with -P option will
   not create them).

2004/01/14:
 - suggested in the README file to store all needed file in a web
   directory in the root of the CVS repository (if the user wants to
   use CVS).

 - moved distributed archive from site-template.tgz to web.tgz.

2004/01/15:
 - Added this history file in the distributed file.
 - Added .cvsignore as suggested by Michael Giroux to avoid dist and
   common dirs to be added in CVS.

2004/03/18:
 - Modify DOCTYPE of project.xml and index.xml to include
   a PUBLIC identifier [Michael Giroux].

 - Modify build.xml to include xmlcatalog to specify location
   of DTD files [Michael Giroux].

 - Modify build.xml xmlvalidate task in the check target to use   
   the xmlcatalog [Michael Giroux].

 - Modify build.xml style task in the real_html target to use
   the xmlcatalog [Michael Giroux].

 - Modified common mini icon names to favicon.ico.

2004/07/09:
 - Some cosmetic changes (reduced header space, removed color hoover
   in a.active links to avoid bad effect with some browsers).

2004/08/09:
 - Google search engine is now used instead of htdig.

 - Remplaced wrong '_new' target by '_blank'.

2005/02/11:
 - Updated copyright mention (2004 -> 2005)

2006/06/08
 - Updated copyright mention (2005 -> 2006)


src/org/objectweb/wildcat/AbstractPath.java
/*
 * Copyright (c) 2004-2005 Universite de Nantes (LINA)
 * Copyright (c) 2005-2006 France Telecom
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Contact: Pierre-Charles David <pcdavid@gmail.com>
 */
package org.objectweb.wildcat;

/**
 * Provides standard implementations for some {@link Path} methods, in particular
 * convenience methods which are defined in terms of others (e.g.
 * {@link Path#isAbsolute()} defined as the opposite of {@link Path#isRelative()}).
 * 
 * @author Pierre-Charles David <pcdavid@gmail.com>
 */
public abstract class AbstractPath implements Path {
    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#isRelative()
     */
    public boolean isRelative() {
        return !isAbsolute();
    }

    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#isPattern()
     */
    public boolean isPattern() {
        return !isDefinite();
    }

    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#isResource()
     */
    public boolean isResource() {
        return !isAttribute();
    }

    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#append(java.lang.String)
     */
    public Path append(String suffix) throws MalformedPathException {
        return append(newFromString(suffix));
    }

    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#prepend(java.lang.String)
     */
    public Path prepend(String prefix) throws MalformedPathException {
        return newFromString(prefix).append(this);
    }

    /**
     * Creates a new path from a string. This method must be implemented by concrete
     * subclasses to benefit from the shortcut methods provided by this class which take a
     * string instead of a fully-formed {@link Path} object.
     */
    protected abstract Path newFromString(String path) throws MalformedPathException;

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#equals(java.lang.Object)
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        } else if (obj instanceof Path) {
            Path that = (Path) obj;
            return (this.isAbsolute() == that.isAbsolute())
                    && (this.getResourcesPart().equals(that.getResourcesPart()))
                    && equals(this.getAttributePart(), that.getAttributePart());
        } else {
            return false;
        }
    }

    /**
     * Tests two objects for equality, correctly handling the case when they are both
     * <code>null</code>.
     * 
     * @param o1
     *            an object
     * @param o2
     *            another object
     * @return <code>true</code> iff <code>o1</code> and <code>o2</code> are
     *         {@link Object#equals(Object)} or both <code>null</code>.
     */
    protected boolean equals(Object o1, Object o2) {
        if (o1 == null) {
            return o2 == null;
        } else {
            return o1.equals(o2);
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#hashCode()
     */
    @Override
    public int hashCode() {
        int hash = isAbsolute() ? 5 : 7;
        for (String step : getResourcesPart()) {
            hash = hash * 13 + step.hashCode();
        }
        String attr = getAttributePart();
        if (attr != null) {
            hash = hash * attr.hashCode();
        }
        return hash;
    }

    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#startsWith(java.lang.String)
     */
    public boolean startsWith(String prefix) throws MalformedPathException {
        return startsWith(newFromString(prefix));
    }

    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#endsWith(java.lang.String)
     */
    public boolean endsWith(String suffix) throws MalformedPathException {
        return endsWith(newFromString(suffix));
    }

    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#subPath(int)
     */
    public Path subPath(int start) throws MalformedPathException {
        return subPath(start, size() + 1);
    }

    /*
     * (non-Javadoc)
     * 
     * @see org.objectweb.wildcat.Path#relativeTo(org.objectweb.wildcat.Path)
     */
    public Path relativeTo(Path base) throws MalformedPathException {
        if (this.startsWith(base)) {
            return this.subPath(base.size() + 1);
        } else {
            throw new MalformedPathException(base + " is not a prefix of " + this + ".");
        }
    }


    /*
     * (non-Javadoc)
     * @see java.lang.Object#toString()
     */
    @Override
    public String toString() {
        StringBuilder str = new StringBuilder();
        boolean first = true;
        if (isAbsolute() && getResourcesPart().size() == 0) {
            return "/";
        }
        for (String step : getResourcesPart()) {
            if (!first || isAbsolute()) {
                str.append("/");
            }
            str.append(step);
            first = false;
        }
        String attr = getAttributePart();
        if (attr != null) {
            str.append("#").append(attr);
        }
        return str.toString();
    }
}
